### ISSUE TITLE
USB PRO: Warnings for "normal" changed packages?

### ISSUE THREAD

**Issue by rewolff** [OpenLightingProject/ola#1290]
I'm implementing an USB_PRO compatible device, reading the code that I'll have to work with... 

https://github.com/OpenLightingProject/ola/blob/master/plugins/usbpro/GenericUsbProWidget.cpp#L273

Here a "changed" packet is handled. The format is designed to be able to compress changes by only sending the changed bytes not all bytes. 

Now this implementation in OLA, that issues a warning if the packet is not the full maximum of 46 bytes, hints at that the original implementation was faulty. 

Suppose DMX channel 30 and 160 have changed. I'd expect the device to send: 
```
<start-of-packet: 7e><packet label: changed-data>  <length: bond, james: 0x007>
  <3: first change is at 3*8=24>  <0x40: the 6th byte has changes> <0x00: no changes in 32-39> <0> <0> <0> <newvalue for ch 30> <39 slots not sent> <end-of-packet: e7>
```
And then another 
```
<length: 7>
  <20: first change is at 20*8=160>  <0x01: the frist byte has changes> <0x00: no more changes> <0> <0> <0> <newvalue for ch 160> <39 slots not sent>
```

I don't HAVE any original devices, so I can't test them. But I'd expect the olad console to be flooded with these warnings for devices doing "the right thing". 

So... IMHO the warning and return should go. Wait! Actually...  The check should be against the minimum packet length of 6 or 7. (minimum valid packet length is 6. But it'd be odd to send one. And "return" is a valid action. So I can live with a warning in that case. The realistic minimum is 7. 

Agreed? Anybody?

Actually... I just realized, you should check agianst BIGGER than the maximum changed-data-packet as well. 
```
if (length > sizeof(widget_data_changed)) {
    OLA_WARN << "Change of state packet was too big: " << length;
    return;
  }
```

**Comment by nomis52** [comment]
It's been a while, but the specification at https://dol2kh495zr52.cloudfront.net/pdf/misc/dmx_usb_pro_api_spec.pdf 

It's not clear on how this is to be interpreted, one way to read it is that the data field varies from 1 to 40 bytes another is that the data field is always 40 bytes long.

Given that I've never seen this warning, I suspect that the official pro product does the latter. The check is there to avoid reading past the end of the data array if the packet is small but upper bits of the changed-bit-array are set.

**Comment by rewolff** [comment]
It says: "Size in bytes: 1 to 40".  This means that the total packet size is 1+5 + 1-40. or 7 to 46. Clearly the size of this packet can be less than 46. 
Then it says: "one byte is present for each set bit", i.e. NO byte is present for each cleared bit.

Now, I don't mind some devices/implementations sending more than what is used, but the whole goal of this part of the protocol is to reduce the communication volume, so sending more bytes than necessary is nonsensical. (I suspect that some devices have a real serial link internally, so if that runs at 115200 kbps, you can't  send the full 250kbps 40 fps DMX bandwidth to the host. So compressing the data by knowing a bit about the data involved is a good plan. 

Let me Try to set up an experiment. (I already  moved the borrowed real enttec pro from "near the door" to "near my desk".) :-)

**Comment by rewolff** [comment]
Update: 
I can't get the enttec to receive data. At all. My own clone works. Not worth my time to investigate further. Sorry. Maybe the cable  to the enttec is broken, something like that.

**Comment by rewolff** [comment]
OK. My bad. My transmitter wasn't transmitting. :-( Sorry about that. 

I have a test where I send the current date/time as DMX data. The enttec reports: 
`00000000  7e 05 08 00 00 00 11 07  1f 0a 34 00 e7           |~.........4..|
`
0x11 = 17 from 2017. 7th mont, 31st day, 10 oclock (AM) 0x32=52 minutes and zero seconds. 

After I changed it... it first reported: 
`00000000  7e 09 0b 00 00 3e 00 00  00 00 11 07 1f 0a 34 e7  |~....>........4.|
`
with an 11 byte long packet payload, consisting of the 1 startbyte,  5 bitfields and the five databytes. Surprisingly not listing the last byte that had a 1/60 chance of being zero at the time I switched. 
From then on I get: 
`00000000  7e 09 07 00 00 40 00 00  00 00 01 e7              |~....@......|
`
packets with 7 databytes, only the seconds have changed.

**Comment by rewolff** [comment]
UPDATE: This issue is now not only a "remove annoying message" but a real "it doesn't work" bug.... 

After doing the above tests, I was "on my way" to start working on my compatible hardware. So I stopped the hexdump and allowed ola to retake the device..... I now get messages like: 

plugins/usbpro/EnttecUsbProWidget.cpp:503: Change of state packet was too small: 7

and because that results in the packet not getting processed, the universe doesn't update at all. This happened because I put the device in send changes only mode outside of ola and ola doesn't reset that. The code is there to _almost_ make it work, just this sanity check is wrong.

**Comment by nomis52** [comment]
Please post a patch and either myself or @peternewman will look at it.

**Comment by rewolff** [comment]
In one place I've coded it like this: 
```
  if ((length < 7) || (length > sizeof(widget_data_changed))) {
    OLA_WARN << "Change of state packet invalid: " << length;
    return;
  }

```
In another place I've written: 
```
  if (length < 7) {
    OLA_WARN << "Change of state packet was too small: " << length;
    return;
  }

  if (length > sizeof (widget_data_changed)) {
    OLA_WARN << "Change of state packet was too big: " << length;
    return;
  }

```
The first was in a git repo, so I could extract a patch: http://prive.bitwizard.nl/len_invalid.patch

The file "plugins/usbpro/EnttecUsbProWidget.cpp" also needs to be patched. 

Which version do you prefer? If you want I can create a pull-request with everything patched, if you tell me which one you like best. (as it's not supposed to happen, the user can't do much about it and an expert has enough information from "invalid", my preference would be to have the smaller code footprint)

**Comment by nomis52** [comment]
The former is fine.

**Comment by rewolff** [comment]
You want me to make a git pull? Ok.

**Comment by nomis52** [comment]
Yes please.

**Comment by peternewman** [comment]
Hi @rewolff . Can we have it in a fresh branch that's not your master one please, it's currently disappeared into the bottom of the https://github.com/OpenLightingProject/ola/pull/1161 PR.

**Comment by rewolff** [comment]
OK. Seems to have worked. I created pull #1301 
yeah. My "master" contains some other stuff that is not relevant here. I seem to have succeeded in creating a fresh branch and adding only this one patch. 

I use Ubuntu on my workstations. When I read "README.debian", and follow the compilation instructions: 1) type autoreconf
I get errors:
`   configure.ac:30: error: required file 'config/compile' not found
`
I think I once figured out what to do, but my memory is not what it used to be. So... I haven't even been able to compile-test this.

**Comment by peternewman** [comment]
Do you want to try with "autoreconf -i"?

In the meantime Travis will check it anyway:
https://travis-ci.org/OpenLightingProject/ola/builds/264109441

**Comment by peternewman** [comment]
Thanks @rewolff . This is now in and will appear in 0.11.0 whenever that launches.

Would you mind testing if OLA now handles the diff packets correctly? If so it may be worth us putting this into 0.10.6.

**Comment by peternewman** [comment]
We should probably also add tests for these short and long change of state packets:
https://github.com/OpenLightingProject/ola/blob/master/plugins/usbpro/EnttecUsbProWidgetTest.cpp#L421

**Comment by rewolff** [comment]
I did test this a while ago. That's what triggered my comment https://github.com/OpenLightingProject/ola/issues/1290#issuecomment-319035159

As to the packet 
```
uint8_t change_of_state_data[] = {
    0, 0x38, 0, 0, 0, 0,
    22, 93, 144, 
  };
```
not specifying the rest of the data is the clearer way of writing that COS packet data. The length (a few lines down) should be specified as 6+3 // six header bytes, 3 databytes . 

Also if you're trying to TEST the code, there is a surprising "times 8" in the protocol. The start is aligned to a multiple of 8, allowing the protocol to use these packets anywhere in a universe of up to 2080 (8* 255+40) bytes.... You don't test the start of changes implementation by specifying zero. Also, If I read that correctly, the interpretation of the data is not checked at all. 

My suggestion would be to send/interpret: 
```
uint8_t change_of_state_data[] = {
    1, 0x28, 1, 0, 0, 0,
    22, 93, 144, 
  };
```
and then verify that bytes 11, 13 and 16 have become 22, 93 and 144.  (the 0x38 bitflag specifies 3 consecutive bytes. Some coding errors might work for consecutive bytes, but not when there are holes. Similarly, I've now moved a change into the second byte, catching coding errors that don't increment the bitpattern byte pointer...)
